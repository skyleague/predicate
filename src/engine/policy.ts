import { collect, stack } from '@skyleague/axioms'
import type { IsEmptyObject, Simplify } from '@skyleague/axioms/types'
import packageJSON from '../../package.json' with { type: 'json' }
import type { FactsFomExprs } from './operator.js'
import type { Expression, ExpressionReturnType } from './types.js'

const { version } = packageJSON

export class EvaluationContext {
    public evaluated = new WeakMap<object, unknown>()
    public state: Partial<Record<PropertyKey, unknown>> = {}
    public input: Record<string, unknown> = {}
    public scope: unknown = {}
    public stack: (readonly [EvaluationContext['evaluated'], EvaluationContext['state'], unknown])[]

    public constructor(input: Record<string, unknown>) {
        this.input = input
        this.stack = [[this.evaluated, this.state, this.scope]]
    }

    public evaluate<O>(expr: Expression<O>): O {
        if (this.evaluated.has(expr)) {
            return this.evaluated.get(expr) as O
        }

        const input = expr.dependsOn.map((d) => this.evaluate(d))

        const result = expr.fn(input, this)
        this.evaluated.set(expr, result)
        return result
    }

    public withScope<T>(scope: unknown, fn: () => T) {
        this.push(scope)
        const result = fn()
        this.pop()
        return result
    }

    protected push(scope: unknown) {
        const top = [new WeakMap<object, unknown>(), {}, scope] as const
        this.stack.push(top)
        this.evaluated = top[0]
        this.state = top[1]
        this.scope = top[2]
    }

    protected pop() {
        this.stack.pop()
        const top = this.stack.at(-1) ?? [this.evaluated, this.state, this.scope]
        this.evaluated = top[0]
        this.state = top[1]
        this.scope = top[2]
    }
}

export interface _EvaluationContext {
    evaluated: WeakMap<object, unknown>
    state: Partial<Record<PropertyKey, unknown>>
    stack: [_EvaluationContext['evaluated'], _EvaluationContext['state'], unknown][]
    scope: unknown
    evaluate<T>(expr: Expression<T>): T
    withScope: <T>(y: unknown, fn: () => T) => T
}

function* collapseExpression(root: Expression[], seen = new WeakSet()) {
    const nodes = stack(root)
    for (const n of nodes) {
        if (!seen.has(n)) {
            yield n
            nodes.push(n.dependsOn)
            seen.add(n)
        }
    }
}
type InferFactName<Expr, k> = Expr extends { name: string } ? Expr['name'] : k

type FilterFactExpressions<Facts, K extends keyof Facts = keyof Facts> = Facts[K] extends {
    _type: 'fact'
}
    ? InferFactName<Facts[K], K>
    : never

type _InputFromName<Names extends PropertyKey, Facts> = {
    [K in Names]: ExpressionReturnType<Extract<Facts[keyof Facts], { name: K }>>
}
type _InputFromExpressions<Facts> = _InputFromName<
    FilterFactExpressions<Facts> extends PropertyKey ? FilterFactExpressions<Facts> : never,
    Facts
>

type _FactsFomExprs<Facts> = Facts extends unknown[] ? Facts[number] : Facts
export type InputFromExpressions<Facts extends Record<string, unknown>> = Simplify<
    _InputFromExpressions<{ [K in keyof Facts as FilterFactExpressions<Facts, K>]: Facts[K] }> &
        _InputFromExpressions<{ [K in keyof Facts]: _FactsFomExprs<FactsFomExprs<Facts[K]>> }>
>

export type OutputFromFacts<Facts extends Record<string, unknown>> = Simplify<{
    [k in keyof Facts]: ExpressionReturnType<Facts[k]>
}>

export interface Policy<I, O> {
    evaluate: IsEmptyObject<I> extends true ? () => { input: I; output: O } : (x: I) => { input: I; output: O }
    expr: () => unknown
}

export function $policy<Facts extends Record<string, Expression>>(
    expressions: Facts,
): Policy<InputFromExpressions<Facts>, OutputFromFacts<Facts>> {
    const facts = Object.entries(expressions).map(([name, e]) => {
        if (e.name === undefined) {
            return Object.assign(e, { name })
        }
        return e
    })

    const allNodes = collect(collapseExpression(facts))
    const inputNodes = allNodes.filter(
        (f: (typeof allNodes)[number]): f is (typeof allNodes)[number] & { name: string } =>
            f._type === 'fact' && f.name !== undefined,
    )
    const outputNodes = allNodes.filter((f) => f._type !== 'fact' && f.name !== undefined)

    const properties = Object.fromEntries(inputNodes.map((e) => [e.name, e.expr('definition')]))
    const outputExpression = Object.fromEntries(outputNodes.map((f) => [f.name, f.expr('expression')]))
    return {
        evaluate: ((input: Record<string, unknown>) => {
            const ctx = new EvaluationContext(input)

            for (const fact of facts) {
                const result = ctx.evaluate(fact)
                if (fact.name !== undefined) {
                    ctx.state[fact.name] = result
                }
            }

            return { input: input, output: ctx.state }
        }) as Policy<InputFromExpressions<Facts>, OutputFromFacts<Facts>>['evaluate'],
        expr: () => {
            const input = properties
            return {
                meta: {
                    version,
                },
                input,
                output: outputExpression,
            }
        },
    }
}
