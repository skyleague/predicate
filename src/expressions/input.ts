import type {
    FactExpression,
    DefinitionType,
    Expression,
    InputExpression,
    InferExpressionType,
    LiteralExpression,
} from '../engine/types.js'
import type { FromExpr } from '../json/jsonexpr.type.js'

import { type JSONPathValue, JSONPath } from '@skyleague/jsonpath'
import type { Schema } from '@skyleague/therefore'

import { inspect } from 'node:util'

export interface Fact<T, Name extends string> extends FactExpression<T> {
    name: Name
    dependsOn: []
    _type: 'fact'
}

export function $fact<T, Name extends string>(schema: Pick<Schema<T>, 'schema' | 'is'>, name: Name): Fact<T, Name> {
    return {
        _type: 'fact',
        dependsOn: [],
        name: name,
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => {
            return ctx.input[name]
        }) as Fact<T, Name>['fn'],
        expr: (definition: DefinitionType) =>
            (definition === 'definition' ? schema.schema : { fact: name.toString() }) as InferExpressionType<T>,
        [inspect.custom]() {
            return `$fact({schema: ${JSON.stringify(schema.schema)}}, "${name}")`
        },
    }
}
export interface From<I, O, DependsOn extends Expression[]> extends InputExpression<O, I, DependsOn, FromExpr> {
    _type: 'value'
}

export function $from<T, Name extends string>(fact: Fact<T, Name>): From<T, JSONPathValue<T, '$'>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]>
export function $from<T, P extends string, Name extends string>(
    fact: Fact<T, Name>,
    path: P = '$' as P
): From<T, JSONPathValue<T, P>, [Fact<T, Name>]> {
    return {
        _type: 'value',
        dependsOn: [fact],
        fn: ((_: unknown, ctx: { input: Record<Name, T> }) => JSONPath.get(ctx.input[fact.name], path)) as From<
            T,
            JSONPathValue<T, P>,
            [Fact<T, Name>]
        >['fn'],
        expr: (_definition) => ({ from: [fact.name, path.toString()] as const }),
        [inspect.custom]() {
            return `$from(${fact[inspect.custom]?.() ?? `"${fact.name}"`}${path === '$' ? '' : `, "${path}"`})`
        },
    }
}

export function $literal<O>(x: O): LiteralExpression<O> {
    return {
        fn: () => x,
        dependsOn: [],
        expr: () => x as InferExpressionType<O>,
        _type: 'literal',
        [inspect.custom]() {
            return `$literal(${inspect(x)})`
        },
    }
}
