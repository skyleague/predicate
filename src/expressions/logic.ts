import { operator } from '../engine/operator.js'
import {
    fromLiteral,
    type Expression,
    type ValueExpression,
    type AsExpression,
    type LiteralOr,
    type ExpressionTypeOfLiteral,
} from '../engine/types.js'
import type { IfExpr } from '../json/jsonexpr.type.js'

export const $if = Object.assign(
    function <CE extends Expression<boolean> | boolean, UE extends LiteralOr<any>, VE extends LiteralOr<any>>(
        _condition: CE,
        _a: UE,
        _b: VE
    ): ValueExpression<
        ExpressionTypeOfLiteral<UE> | ExpressionTypeOfLiteral<VE>,
        [AsExpression<CE>, AsExpression<UE>, AsExpression<VE>],
        IfExpr
    > {
        const condition = fromLiteral(_condition)
        const a = fromLiteral(_a)
        const b = fromLiteral(_b)
        return {
            fn: ([conditional, x1, x2]) => (conditional ? x1 : x2) as ExpressionTypeOfLiteral<UE> | ExpressionTypeOfLiteral<VE>,
            dependsOn: [condition, a, b],
            expr: (mod) => ({
                if: [condition.expr(mod), a.expr(mod), b.expr(mod)],
            }),
        }
    },
    { operator: 'if', symbol: '$if' }
)

export const $and = operator({ operator: 'and', symbol: '$and', fn: (xs: boolean[]) => xs.every((x) => x) })

export const $or = operator({ operator: 'or', symbol: '$or', fn: (xs: boolean[]) => xs.some((x) => x) })
