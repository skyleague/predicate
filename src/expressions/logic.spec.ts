import { $fact, $from, type Fact, type From } from './input.js'
import { $and, $if, $or } from './logic.js'

import { LogicObj } from '../../test/logic.type.js'
import { $policy } from '../engine/policy.js'
import type { LiteralExpression, ValueExpression } from '../engine/types.js'
import type { AndExpr, BooleanExpr, IfExpr, NumberExpr, OrExpr, StringExpr } from '../json/jsonexpr.type.js'

import { forAll } from '@skyleague/axioms'
import { arbitrary } from '@skyleague/therefore'
import { describe, expect, expectTypeOf, it } from 'vitest'

describe('if', () => {
    it('handles a simple schema case', () => {
        const input = {
            c: true,
            a: 1,
            b: 2,
            d: [],
            e: true,
        }
        const fact = $fact(LogicObj, 'input')
        const c = $from(fact, '$.c')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $if(c, x2, 1)
        const fn = $policy({ x1, x2, x3 })
        const result = fn.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": 1,
                "b": 2,
                "c": true,
                "d": [],
                "e": true,
              },
            },
            "output": {
              "x1": 1,
              "x2": 2,
              "x3": 2,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ x1: LogicObj['a']; x2: LogicObj['a']; x3: LogicObj['a'] }>()
        expect(fn.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(LogicObj, 'input')
        const c = $from(fact, '$.c')
        const x1 = $from(fact, '$.a')
        const x2 = $from(fact, '$.b')
        const x3 = $if(c, x1, x2)
        const fn = $policy({ x1, x2, x3 })
        forAll(arbitrary(LogicObj), (input) => {
            const result = fn.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({ x1: input.a, x2: input.b, x3: input.c ? input.a : input.b })
            expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
            expectTypeOf(result.output).toEqualTypeOf<{ x1: LogicObj['a']; x2: LogicObj['a']; x3: LogicObj['a'] }>()
        })
    })

    it('handles the types', () => {
        const x1 = $if(false, 1, 2)
        expectTypeOf(x1).toEqualTypeOf<
            ValueExpression<
                number,
                [
                    LiteralExpression<false, BooleanExpr>,
                    LiteralExpression<number, NumberExpr>,
                    LiteralExpression<number, NumberExpr>,
                ],
                never,
                IfExpr
            >
        >()

        const x2 = $if(true, 1, '2')
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<
                string | number,
                [
                    LiteralExpression<true, BooleanExpr>,
                    LiteralExpression<number, NumberExpr>,
                    LiteralExpression<string, StringExpr>,
                ],
                never,
                IfExpr
            >
        >()
    })
})

describe('and', () => {
    it('handles a simple schema case', () => {
        const input = {
            c: true,
            a: 1,
            b: 2,
            d: [],
            e: true,
        }
        const fact = $fact(LogicObj, 'input')
        const c = $from(fact, '$.c')
        const e = $from(fact, '$.e')
        const x2 = $and(c, e)
        const fn = $policy({ c, e, x2 })
        const result = fn.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": 1,
                "b": 2,
                "c": true,
                "d": [],
                "e": true,
              },
            },
            "output": {
              "c": true,
              "e": true,
              "x2": true,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ c: boolean; e: boolean; x2: boolean }>()
        expect(fn.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(LogicObj, 'input')
        const c = $from(fact, '$.c')
        const e = $from(fact, '$.e')
        const x2 = $and(c, e)
        const fn = $policy({ c, e, x2 })
        forAll(arbitrary(LogicObj), (input) => {
            const result = fn.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({ c: input.c, e: input.e, x2: input.c && input.e })
            expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
            expectTypeOf(result.output).toEqualTypeOf<{ c: boolean; e: boolean; x2: boolean }>()
        })
    })

    it('handles the types', () => {
        const fact = $fact(LogicObj, 'input')

        const x1 = $and(false, true, false)
        expectTypeOf(x1).toEqualTypeOf<
            ValueExpression<
                boolean,
                (LiteralExpression<false, BooleanExpr> | LiteralExpression<true, BooleanExpr>)[],
                never,
                AndExpr
            >
        >()

        const x2 = $and($from(fact, '$.c'), true)
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<
                boolean,
                (LiteralExpression<true, BooleanExpr> | From<LogicObj, boolean, Fact<LogicObj, 'input'>>)[],
                [Fact<LogicObj, 'input'>],
                AndExpr
            >
        >()

        const x3 = $and(true, $from(fact, '$.e'))
        expectTypeOf(x3).toEqualTypeOf<
            ValueExpression<
                boolean,
                (From<LogicObj, boolean, Fact<LogicObj, 'input'>> | LiteralExpression<true, BooleanExpr>)[],
                [Fact<LogicObj, 'input'>],
                AndExpr
            >
        >()
    })
})

describe('or', () => {
    it('handles a simple schema case', () => {
        const input = {
            c: true,
            a: 1,
            b: 2,
            d: [],
            e: true,
        }
        const fact = $fact(LogicObj, 'input')
        const c = $from(fact, '$.c')
        const e = $from(fact, '$.e')
        const x2 = $or(c, e)
        const fn = $policy({ c, e, x2 })
        const result = fn.evaluate({ input })

        expect(result).toMatchInlineSnapshot(`
          {
            "input": {
              "input": {
                "a": 1,
                "b": 2,
                "c": true,
                "d": [],
                "e": true,
              },
            },
            "output": {
              "c": true,
              "e": true,
              "x2": true,
            },
          }
        `)
        expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
        expectTypeOf(result.output).toEqualTypeOf<{ c: boolean; e: boolean; x2: boolean }>()
        expect(fn.expr()).toMatchSnapshot()
    })

    it('handles any simple schema case', () => {
        const fact = $fact(LogicObj, 'input')
        const c = $from(fact, '$.c')
        const e = $from(fact, '$.e')
        const x2 = $or(c, e)
        const fn = $policy({ c, e, x2 })
        forAll(arbitrary(LogicObj), (input) => {
            const result = fn.evaluate({ input })
            expect(result.input).toEqual({ input })
            expect(result.output).toEqual({ c: input.c, e: input.e, x2: input.c || input.e })
            expectTypeOf(result.input.input).toEqualTypeOf<LogicObj>()
            expectTypeOf(result.output).toEqualTypeOf<{ c: boolean; e: boolean; x2: boolean }>()
        })
    })

    it('handles the types', () => {
        const fact = $fact(LogicObj, 'input')

        const x1 = $or(false, true, false)
        expectTypeOf(x1).toEqualTypeOf<
            ValueExpression<
                boolean,
                (LiteralExpression<false, BooleanExpr> | LiteralExpression<true, BooleanExpr>)[],
                never,
                OrExpr
            >
        >()

        const x2 = $or($from(fact, '$.c'), true)
        expectTypeOf(x2).toEqualTypeOf<
            ValueExpression<
                boolean,
                (From<LogicObj, boolean, Fact<LogicObj, 'input'>> | LiteralExpression<true, BooleanExpr>)[],
                [Fact<LogicObj, 'input'>],
                OrExpr
            >
        >()

        const x3 = $or(true, $from(fact, '$.e'))
        expectTypeOf(x3).toEqualTypeOf<
            ValueExpression<
                boolean,
                (From<LogicObj, boolean, Fact<LogicObj, 'input'>> | LiteralExpression<true, BooleanExpr>)[],
                [Fact<LogicObj, 'input'>],
                OrExpr
            >
        >()
    })
})
