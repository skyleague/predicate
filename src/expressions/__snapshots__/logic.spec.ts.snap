// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`and > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "b": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "c": {
          "type": "boolean",
        },
        "d": {
          "items": {
            "additionalProperties": true,
            "properties": {
              "a": {
                "type": "boolean",
              },
              "b": {
                "type": "boolean",
              },
            },
            "required": [
              "a",
              "b",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "e": {
          "type": "boolean",
        },
      },
      "required": [
        "c",
        "a",
        "b",
        "d",
        "e",
      ],
      "title": "LogicObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "c": {
      "from": [
        "input",
        "$.c",
      ],
    },
    "e": {
      "from": [
        "input",
        "$.e",
      ],
    },
    "x2": {
      "and": [
        {
          "from": [
            "input",
            "$.c",
          ],
        },
        {
          "from": [
            "input",
            "$.e",
          ],
        },
      ],
    },
  },
}
`;

exports[`if > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "b": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "c": {
          "type": "boolean",
        },
        "d": {
          "items": {
            "additionalProperties": true,
            "properties": {
              "a": {
                "type": "boolean",
              },
              "b": {
                "type": "boolean",
              },
            },
            "required": [
              "a",
              "b",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "e": {
          "type": "boolean",
        },
      },
      "required": [
        "c",
        "a",
        "b",
        "d",
        "e",
      ],
      "title": "LogicObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "x1": {
      "from": [
        "input",
        "$.a",
      ],
    },
    "x2": {
      "from": [
        "input",
        "$.b",
      ],
    },
    "x3": {
      "if": [
        {
          "from": [
            "input",
            "$.c",
          ],
        },
        {
          "from": [
            "input",
            "$.b",
          ],
        },
        1,
      ],
    },
  },
}
`;

exports[`or > handles a simple schema case 2`] = `
{
  "input": {
    "input": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "a": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "b": {
          "anyOf": [
            {
              "type": "string",
            },
            {
              "type": "boolean",
            },
            {
              "type": "integer",
            },
          ],
        },
        "c": {
          "type": "boolean",
        },
        "d": {
          "items": {
            "additionalProperties": true,
            "properties": {
              "a": {
                "type": "boolean",
              },
              "b": {
                "type": "boolean",
              },
            },
            "required": [
              "a",
              "b",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "e": {
          "type": "boolean",
        },
      },
      "required": [
        "c",
        "a",
        "b",
        "d",
        "e",
      ],
      "title": "LogicObj",
      "type": "object",
    },
  },
  "meta": {
    "version": "1.0.0",
  },
  "output": {
    "c": {
      "from": [
        "input",
        "$.c",
      ],
    },
    "e": {
      "from": [
        "input",
        "$.e",
      ],
    },
    "x2": {
      "or": [
        {
          "from": [
            "input",
            "$.c",
          ],
        },
        {
          "from": [
            "input",
            "$.e",
          ],
        },
      ],
    },
  },
}
`;
